src/lib/ptree/json.lua

```diff
@@ -140,13 +208,44 @@ function write_json_object(output, obj)
    local comma = false
    for k,v in pairs(obj) do
       if comma then output:write_chars(',') else comma = true end
-      write_json_string(output, k)
+      write_json_string(output, k) -- JSON keys must be strings
       output:write_chars(':')
-      write_json_string(output, v)
+      write_json_type(output, v)
    end
    output:write_chars('}')
 end
 
+local function write_json_array(output, obj)
+   output:write_chars('[')
+   for i,v in ipairs(obj) do
+      if i > 1 then output:write_chars(',') end
+      write_json_type(output, v)
+   end
+   output:write_chars(']')
+end
+
+write_json_type = function(output, var)
+   local tp = type(var)
+   if var == NIL then
+      return write_json_null(output)
+   end
+   if tp == 'table' then
```

Comment:
Can you use elseif here, and add a comment that the NIL case has to be first because it's represented as a table?